src/Util/MIMETypeUtil.m

```diff
@@ -277,7 +277,65 @@ + (BOOL)isAudio:(NSString *)contentType {
 
 + (NSString *)filePathForAttachment:(NSString *)uniqueId
                          ofMIMEType:(NSString *)contentType
-                           inFolder:(NSString *)folder {
+                           filename:(nullable NSString *)filename
+                           inFolder:(NSString *)folder
+{
+    NSString *kDefaultFileExtension = @"bin";
+
+    if (filename.length > 0) {
+        NSString *normalizedFilename =
+            [filename stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
+        // Ensure that the filename is a valid filesystem name,
+        // replacing invalid characters with an underscore.
+        for (NSCharacterSet *invalidCharacterSet in @[
+                 [NSCharacterSet whitespaceCharacterSet],
+                 [NSCharacterSet newlineCharacterSet],
+                 [NSCharacterSet illegalCharacterSet],
+                 [NSCharacterSet controlCharacterSet],
+                 [NSCharacterSet characterSetWithCharactersInString:@"<>|\\:()&;?*"],
```

Comment:
I've added forward-slash / and tilde ~, but I'm omitting period since it's so widely used in filenames like: "Budget.latest.xls" or whatever.  I'll do explicit checks for filenames like "." and "..".